core/src/main/java/org/elasticsearch/search/SearchHit.java

```diff
@@ -465,129 +473,166 @@ public XContentBuilder toInnerXContent(XContentBuilder builder, Params params) t
         if (shardId != null && nodeId != null) {
             searchHit.shard(new SearchShardTarget(nodeId, shardId));
         }
         searchHit.fields(fields);
         return searchHit;
     }
 
-    private static Explanation parseExplanation(XContentParser parser) throws IOException {
+    private static <T> T get(String key, Map<String, Object> map, T defaultValue) {
+        @SuppressWarnings("unchecked")
+        T value = (T) map.get(key);
+        if (value == null) {
+            value = defaultValue;
+        }
+        return value;
+    }
+
+    private static float parseScore(XContentParser parser, Void context) throws IOException {
+        if (parser.currentToken() == XContentParser.Token.VALUE_NUMBER || parser.currentToken() == XContentParser.Token.VALUE_STRING) {
+            return parser.floatValue();
+        } else {
+            return Float.NaN;
+        }
+    }
+
+    private static BytesReference parseSourceBytes(XContentParser parser, Void context) throws IOException {
+        try (XContentBuilder builder = XContentBuilder.builder(parser.contentType().xContent())) {
+            // the original document gets slightly modified: whitespaces or
+            // pretty printing are not preserved,
+            // it all depends on the current builder settings
+            builder.copyCurrentStructure(parser);
+            return builder.bytes();
+        }
+    }
+
+    /**
+     * we need to declare parse fields for each metadata field, except for _ID, _INDEX and _TYPE which are
+     * handled individually. All other fields are parsed to an entry in the fields map
+     */
+    private static void declareMetaDataFields(ObjectParser<Map<String, Object>, Void> parser) {
+        for (String metadatafield : MapperService.getAllMetaFields()) {
+            if (metadatafield.equals(Fields._ID) == false && metadatafield.equals(Fields._INDEX) == false
+                    && metadatafield.equals(Fields._TYPE) == false) {
+                parser.declareField((map, field) -> {
+                    @SuppressWarnings("unchecked")
+                    Map<String, SearchHitField> fieldMap = (Map<String, SearchHitField>) map.get(Fields.FIELDS);
+                    if (fieldMap == null) {
+                        fieldMap = new HashMap<>();
+                        map.put(Fields.FIELDS, fieldMap);
+                    }
+                    fieldMap.put(field.getName(), field);
+                }, (p, c) -> {
+                    List<Object> values = new ArrayList<>();
+                    values.add(parseStoredFieldsValue(p));
+                    return new SearchHitField(metadatafield, values);
+                }, new ParseField(metadatafield), ValueType.VALUE);
+            }
+        }
+    }
+
+    private static Map<String, SearchHitField> parseFields(XContentParser parser, Void context) throws IOException {
+        Map<String, SearchHitField> fields = new HashMap<>();
```

Comment:
Right... but I'd be happy if we could unit test this, and if we do then we need to ensure the object start.